package/kernel/mac80211/patches/subsys/200-mac80211-introduce-support-for-vendor-QAM-256-in-2.4.patch

```diff
@@ -0,0 +1,230 @@
+From be3fd3c999fda3b419da93a659b86e99ccf5e214 Mon Sep 17 00:00:00 2001
+From: Christian Marangi <ansuelsmth@gmail.com>
+Date: Wed, 14 Jun 2023 07:15:48 +0200
+Subject: [PATCH 1/4] mac80211: introduce support for vendor QAM-256 in 2.4GHz
+ 802.11n
+
+Some vendor supports non-standard QAM-256 in 2.4GHz 802.11n mode.
+The implementation works by comunicating vht capabilities to the client
+in 2.4GHz 802.11n, the supported client will take this info and
+benefits from the additional rates of it.
+
+Each driver needs to enable support for this by enabling the
+vendor_qam256_supported in the 2G sband struct and add the required
+capabilities for vht_cap.
+
+This feature is supported by various vendor with all kind of marketing
+name, but all of them have in common the use of vht capabilities in
+2.4GHz 802.11n.
+
+Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
+---
+ include/net/cfg80211.h |  2 ++
+ net/mac80211/mlme.c    | 12 ++++++----
+ net/mac80211/util.c    | 10 +++++---
+ net/mac80211/vht.c     | 52 +++++++++++++++++++++++++++---------------
+ 4 files changed, 50 insertions(+), 26 deletions(-)
+
+diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
+index 1b8619685bf6..e9c9afae9a80 100644
+--- a/include/net/cfg80211.h
++++ b/include/net/cfg80211.h
+@@ -338,11 +338,13 @@ struct ieee80211_sta_ht_cap {
+  * to describe 802.11ac VHT capabilities for an STA.
+  *
+  * @vht_supported: is VHT supported by the STA
++ * @qam256_supported: is QAM256 supported by the STA
+  * @cap: VHT capabilities map as described in 802.11ac spec
+  * @vht_mcs: Supported VHT MCS rates
+  */
+ struct ieee80211_sta_vht_cap {
+ 	bool vht_supported;
++	bool vendor_qam256_supported;
+ 	u32 cap; /* use IEEE80211_VHT_CAP_ */
+ 	struct ieee80211_vht_mcs_info vht_mcs;
+ };
+diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
+index 3827b5dcdb03..398030f91f48 100644
+--- a/net/mac80211/mlme.c
++++ b/net/mac80211/mlme.c
+@@ -4720,7 +4720,7 @@ static int ieee80211_prep_channel(struct ieee80211_sub_if_data *sdata,
+ 	const struct cfg80211_bss_ies *ies;
+ 	int ret;
+ 	u32 i;
+-	bool have_80mhz;
++	bool can_vht;
+ 
+ 	rcu_read_lock();
+ 
+@@ -4867,17 +4867,21 @@ static int ieee80211_prep_channel(struct ieee80211_sub_if_data *sdata,
+ 	}
+ 
+ 	/* Allow VHT if at least one channel on the sband supports 80 MHz */
+-	have_80mhz = false;
++	can_vht = false;
+ 	for (i = 0; i < sband->n_channels; i++) {
+ 		if (sband->channels[i].flags & (IEEE80211_CHAN_DISABLED |
+ 						IEEE80211_CHAN_NO_80MHZ))
+ 			continue;
+ 
+-		have_80mhz = true;
++		can_vht = true;
+ 		break;
+ 	}
+ 
+-	if (!have_80mhz) {
++	/* Some Vendor supports non-standard QAM-256 on 2.4GHz 802.11n */
++	if (sband->vht_cap.vendor_qam256_supported)
++		can_vht = true;
++
++	if (!can_vht) {
+ 		sdata_info(sdata, "80 MHz not supported, disabling VHT\n");
```

Comment:
This message needs fixing too, right?

Reply:
that will only be triggered for 5GHz.

Reply:
but why? aren't there "wave 1" and "wave 2" devices, where the former cannot support VHT on 2.4?

Reply:
sorry i tought this was on a different codepath. Yes I have to update the message and add entry about vendor QAM-256.

Reply:
I don't think "wave 1" or "wave 2" devices has anything to do with VHT on 2.4 though?

Reply:
@rany2 maybe vendor-speak is to be taken lightly, but my point of reference is that DAP-1620 rev A was advertised as "AC1200", whereas rev B was advertised as "AC1300 Wave 2".